src/scmdata/database.py

```diff
@@ -0,0 +1,266 @@
+existing data.
+
+Subsets of data are also able to be read without having to load all the data
+and then filter. For example, one save model results from a number of different
+climate models and then load just the ``Surface Temperature`` data for all models.
+"""
+import glob
+import os
+import os.path
+import pathlib
+import shutil
+
+import pandas as pd
+import tqdm.autonotebook as tqdman
+
+from scmdata import ScmRun, run_append
+
+
+def ensure_dir_exists(fp):
+    """
+    Ensure directory exists
+
+    Parameters
+    ----------
+    fp : str
+        Filepath of which to ensure the directory exists
+    """
+    dir_to_check = os.path.dirname(fp)
+    if not os.path.isdir(dir_to_check):
+        try:
+            os.makedirs(dir_to_check)
+        except OSError:
+            # Prevent race conditions if multiple threads attempt to create dir at same time
+            if not os.path.exists(dir_to_check):
+                raise
+
+
+def _check_is_subdir(root, d):
+    root_path = pathlib.Path(root).resolve()
+    out_path = pathlib.Path(d).resolve()
+
+    is_subdir = root_path in out_path.parents
+    # Sanity check that we never mangle anything outside of the root dir
+    if not is_subdir:  # pragma: no cover
+        raise AssertionError("{} not in {}".format(d, root))
+
+
+class ScmDatabase:
+    """
+    On-disk database handler for outputs from SCMs
+
+    Data is split into groups as specified by :attr:`levels`. This allows for fast
+    reading and writing of new subsets of data when a single output file is no longer
+    performant or data cannot all fit in memory.
+    """
+
+    def __init__(
+        self, root_dir, levels=("climate_model", "variable", "region", "scenario"),
+    ):
+        """
+        Initialise the database
+
+        Parameters
+        ----------
+        root_dir : str
+            The root directory of the database
+        levels : tuple of str
+            Specifies how the runs should be stored on disk.
+
+            The data will be grouped by ``levels``. These levels should be adapted to
+            best match the input data and desired access pattern. If there are any additional
+            varying dimensions, they will be stored as dimensions.
+
+        .. note::
+
+            Creating a new :class:`ScmDatabase` does not modify any existing data on disk. To
+            load an existing database ensure that the :attr:`root_dir` and :attr:`levels` are
+            the same as the previous instance.
+        """
+        self._root_dir = root_dir
+        self.levels = levels
+
+    def __repr__(self):
+        return "<scmdata.database.SCMDatabase (root_dir: {})>".format(self._root_dir)
+
+    @property
+    def root_dir(self):
+        """
+        Root directory of the database.
+
+        Returns
+        -------
+        str
+        """
+        return self._root_dir
+
+    @staticmethod
+    def _get_disk_filename(inp):
+        return inp.replace("|", "-").replace(" ", "-")
```

Comment:
is it worth adding a few more safety valves here to guard against other things like commas, colons, dollar signs?